compact.go

```diff
@@ -816,17 +812,32 @@ func (c *LeveledCompactor) populateBlock(blocks []BlockReader, meta *BlockMeta,
 			}
 			valset.set(l.Value)
 		}
-		postings.Add(i, lset)
-
-		i++
 	}
 	if set.Err() != nil {
 		return errors.Wrap(set.Err(), "iterate compaction set")
 	}
 
-	s := make([]string, 0, 256)
+	if meta.Stats.NumSamples == 0 {
+		// No postings to write, exit early.
+		return nil
+	}
+
+	var (
+		maxNumValues   int
+		numLabelValues int
+		names          = make([]string, 0, len(values)+1)
+	)
+	for _, v := range values {
```

Comment:
nice, wonder if we can move that to separate function

Reply:
Done

Reply:
Not sure which part you wanted to move to a separate function, I just moved the entire code of writing postings into a separate function as that was getting big.